File: backend/src/test/java/com/equitrack/service/PositionServiceTest.java
package com.equitrack.service;

import com.equitrack.model.Position;
import com.equitrack.model.Trade;
import com.equitrack.model.Transaction;
import com.equitrack.model.TransactionAction;
import com.equitrack.model.TransactionSide;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
public class PositionServiceTest {

    private PositionService positionService;

    @BeforeEach
    void setUp() {
        positionService = new PositionService();
    }

    @Test
    void testSampleDataProcessing() {
        // Sample transactions from the problem statement
        List<Transaction> sampleTransactions = List.of(
            new Transaction(1L, 1L, 1, "REL", 50, TransactionAction.INSERT, TransactionSide.Buy),
            new Transaction(2L, 2L, 1, "ITC", 40, TransactionAction.INSERT, TransactionSide.Sell),
            new Transaction(3L, 3L, 1, "INF", 70, TransactionAction.INSERT, TransactionSide.Buy),
            new Transaction(4L, 1L, 2, "REL", 60, TransactionAction.UPDATE, TransactionSide.Buy),
            new Transaction(5L, 2L, 2, "ITC", 30, TransactionAction.CANCEL, TransactionSide.Buy),
            new Transaction(6L, 4L, 1, "INF", 20, TransactionAction.INSERT, TransactionSide.Sell)
        );

        // Process transactions
        positionService.processTransactions(sampleTransactions);

        // Get final positions
        List<Position> positions = positionService.getPositions();
        List<Trade> trades = positionService.getTrades();

        // Verify expected results
        assertEquals(3, positions.size());

        // Check REL position (should be +60 after UPDATE)
        Position relPosition = positions.stream()
                .filter(p -> "REL".equals(p.getSecurityCode()))
                .findFirst()
                .orElse(null);
        assertNotNull(relPosition);
        assertEquals(60, relPosition.getQuantity());

        // Check ITC position (should be 0 after CANCEL)
        Position itcPosition = positions.stream()
                .filter(p -> "ITC".equals(p.getSecurityCode()))
                .findFirst()
                .orElse(null);
        assertNotNull(itcPosition);
        assertEquals(0, itcPosition.getQuantity());

        // Check INF position (should be +50: 70 Buy - 20 Sell)
        Position infPosition = positions.stream()
                .filter(p -> "INF".equals(p.getSecurityCode()))
                .findFirst()
                .orElse(null);
        assertNotNull(infPosition);
        assertEquals(50, infPosition.getQuantity());

        // Verify trades
        assertEquals(4, trades.size());

        // Check that Trade 2 is cancelled
        Trade cancelledTrade = trades.stream()
                .filter(t -> t.getTradeId().equals(2L))
                .findFirst()
                .orElse(null);
        assertNotNull(cancelledTrade);
        assertTrue(cancelledTrade.getIsCancelled());

        // Check that Trade 1 is updated to version 2
        Trade updatedTrade = trades.stream()
                .filter(t -> t.getTradeId().equals(1L))
                .findFirst()
                .orElse(null);
        assertNotNull(updatedTrade);
        assertEquals(2, updatedTrade.getCurrentVersion());
        assertEquals(60, updatedTrade.getQuantity());
    }

    @Test
    void testInsertTransaction() {
        Transaction transaction = new Transaction(1L, 1L, 1, "AAPL", 100, TransactionAction.INSERT, TransactionSide.Buy);
        positionService.processTransaction(transaction);

        List<Position> positions = positionService.getPositions();
        assertEquals(1, positions.size());
        assertEquals(100, positions.get(0).getQuantity());
        assertEquals("AAPL", positions.get(0).getSecurityCode());
    }

    @Test
    void testUpdateTransaction() {
        // First insert
        Transaction insert = new Transaction(1L, 1L, 1, "AAPL", 100, TransactionAction.INSERT, TransactionSide.Buy);
        positionService.processTransaction(insert);

        // Then update
        Transaction update = new Transaction(2L, 1L, 2, "AAPL", 150, TransactionAction.UPDATE, TransactionSide.Buy);
        positionService.processTransaction(update);

        List<Position> positions = positionService.getPositions();
        assertEquals(1, positions.size());
        assertEquals(150, positions.get(0).getQuantity());
    }

    @Test
    void testCancelTransaction() {
        // First insert
        Transaction insert = new Transaction(1L, 1L, 1, "AAPL", 100, TransactionAction.INSERT, TransactionSide.Buy);
        positionService.processTransaction(insert);

        // Then cancel
        Transaction cancel = new Transaction(2L, 1L, 2, "AAPL", 100, TransactionAction.CANCEL, TransactionSide.Buy);
        positionService.processTransaction(cancel);

        List<Position> positions = positionService.getPositions();
        assertEquals(1, positions.size());
        assertEquals(0, positions.get(0).getQuantity());
    }

    @Test
    void testMixedBuySell() {
        Transaction buy = new Transaction(1L, 1L, 1, "AAPL", 100, TransactionAction.INSERT, TransactionSide.Buy);
        Transaction sell = new Transaction(2L, 2L, 1, "AAPL", 30, TransactionAction.INSERT, TransactionSide.Sell);

        positionService.processTransaction(buy);
        positionService.processTransaction(sell);

        List<Position> positions = positionService.getPositions();
        assertEquals(1, positions.size());
        assertEquals(70, positions.get(0).getQuantity()); // 100 - 30
    }

    @Test
    void testClearData() {
        Transaction transaction = new Transaction(1L, 1L, 1, "AAPL", 100, TransactionAction.INSERT, TransactionSide.Buy);
        positionService.processTransaction(transaction);

        assertEquals(1, positionService.getPositions().size());

        positionService.clear();

        assertEquals(0, positionService.getPositions().size());
        assertEquals(0, positionService.getTrades().size());
    }
}

File: backend/src/test/PositionService.test.ts
import { PositionService } from '../services/PositionService';
import { Transaction } from '../types';

// Test the business logic with the sample data
function testPositionService() {
  console.log('🧪 Testing Position Service...\n');

  const positionService = new PositionService();

  // Sample transactions from the problem statement
  const sampleTransactions: Transaction[] = [
    { transactionId: 1, tradeId: 1, version: 1, securityCode: 'REL', quantity: 50, action: 'INSERT', side: 'Buy' },
    { transactionId: 2, tradeId: 2, version: 1, securityCode: 'ITC', quantity: 40, action: 'INSERT', side: 'Sell' },
    { transactionId: 3, tradeId: 3, version: 1, securityCode: 'INF', quantity: 70, action: 'INSERT', side: 'Buy' },
    { transactionId: 4, tradeId: 1, version: 2, securityCode: 'REL', quantity: 60, action: 'UPDATE', side: 'Buy' },
    { transactionId: 5, tradeId: 2, version: 2, securityCode: 'ITC', quantity: 30, action: 'CANCEL', side: 'Buy' },
    { transactionId: 6, tradeId: 4, version: 1, securityCode: 'INF', quantity: 20, action: 'INSERT', side: 'Sell' }
  ];

  console.log('📊 Processing sample transactions...\n');

  // Process transactions one by one to show intermediate states
  sampleTransactions.forEach((transaction, index) => {
    console.log(`Transaction ${index + 1}:`, {
      tradeId: transaction.tradeId,
      version: transaction.version,
      security: transaction.securityCode,
      quantity: transaction.quantity,
      action: transaction.action,
      side: transaction.side
    });

    positionService.processTransaction(transaction);

    const positions = positionService.getPositions();
    console.log('Current positions:', positions);
    console.log('---');
  });

  // Final positions
  const finalPositions = positionService.getPositions();
  console.log('\n🎯 Final Positions:');
  finalPositions.forEach(position => {
    const sign = position.quantity > 0 ? '+' : '';
    console.log(`${position.securityCode}: ${sign}${position.quantity}`);
  });

  // Verify expected results
  console.log('\n✅ Expected Results:');
  console.log('REL: +60 (Updated from 50 to 60)');
  console.log('ITC: 0 (Cancelled trade)');
  console.log('INF: +50 (70 Buy - 20 Sell)');

  // Get all trades
  const trades = positionService.getTrades();
  console.log('\n📋 All Trades:');
  trades.forEach(trade => {
    console.log(`Trade ${trade.tradeId}: ${trade.securityCode} ${trade.quantity} ${trade.side} - ${trade.isCancelled ? 'CANCELLED' : 'ACTIVE'} (v${trade.currentVersion})`);
  });

  return finalPositions;
}

// Run the test if this file is executed directly
if (require.main === module) {
  testPositionService();
}

export { testPositionService };
